File: 백준/Gold/2234. 성곽/성곽.java
import java.util.Scanner;

public class Main {
    static int n, m, mx, big, cnt;
    static int[][] visited;
    static int[][] a;
    static int[] compSize;
    static final int[] dy = {0, -1, 0, 1};
    static final int[] dx = {-1, 0, 1, 0};

    static int dfs(int y, int x, int cnt) {
        if (visited[y][x] != 0) {
            return 0;
        }
        visited[y][x] = cnt;
        int ret = 1;
        for (int i = 0; i < 4; i++) {
            if ((a[y][x] & (1 << i)) == 0) {
                int ny = y + dy[i];
                int nx = x + dx[i];
                ret += dfs(ny, nx, cnt);
            }
        }
        return ret;
    }

    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        n = scanner.nextInt();
        m = scanner.nextInt();
        visited = new int[m][n];
        a = new int[m][n];
        compSize = new int[2504];

        for (int i = 0; i < m; i++) {
            for (int j = 0; j < n; j++) {
                a[i][j] = scanner.nextInt();
            }
        }

        for (int i = 0; i < m; i++) {
            for (int j = 0; j < n; j++) {
                if (visited[i][j] == 0) {
                    cnt++;
                    compSize[cnt] = dfs(i, j, cnt);
                    mx = Math.max(mx, compSize[cnt]);
                }
            }
        }

        for (int i = 0; i < m; i++) {
            for (int j = 0; j < n; j++) {
                if (i + 1 < m) {
                    int a = visited[i + 1][j];
                    int b = visited[i][j];
                    if (a != b) {
                        big = Math.max(big, compSize[a] + compSize[b]);
                    }
                }
                if (j + 1 < n) {
                    int a = visited[i][j + 1];
                    int b = visited[i][j];
                    if (a != b) {
                        big = Math.max(big, compSize[a] + compSize[b]);
                    }
                }
            }
        }

        System.out.println(cnt);
        System.out.println(mx);
        System.out.println(big);
    }
}

File: 백준/Gold/2234. 성곽/성곽.py
import sys

input = sys.stdin.readline

dy, dx = [0, -1, 0, 1], [-1, 0, 1, 0]

def dfs(y, x, cnt):
    if visited[y][x]: return 0
    visited[y][x] = cnt
    size = 1
    for i in range(4):
        if not (grid[y][x] & (1 << i)):
            ny, nx = y + dy[i], x + dx[i]
            if 0 <= ny < m and 0 <= nx < n:
                size += dfs(ny, nx, cnt)
    return size

n, m = map(int, input().split())
grid = [list(map(int, input().split())) for _ in range(m)]
visited = [[0]*n for _ in range(m)]
comp_size = [0]*(m*n+1)

cnt = 0
mx = 0
for i in range(m):
    for j in range(n):
        if not visited[i][j]:
            cnt += 1
            comp_size[cnt] = dfs(i, j, cnt)
            mx = max(mx, comp_size[cnt])

big = 0
for i in range(m):
    for j in range(n):
        # 남쪽 체크
        if i + 1 < m:
            a, b = visited[i + 1][j], visited[i][j]
            if a != b:
                big = max(big, comp_size[a] + comp_size[b])

        #동쪽 체크
        if j + 1 < n:
            a, b = visited[i][j + 1], visited[i][j]
            if a != b:
                big = max(big, comp_size[a] + comp_size[b])
print(cnt, mx, big, sep = '\n')